src/KDMqtt/mqtt.h

```diff
@@ -0,0 +1,250 @@
+    bool m_isInitialized;
+};
+
+/*
+ * Class: IMqttClient
+ *
+ * This is an abstract class specifing a generic
+ * interface exposed to application / business logic
+ * to access MQTT client implementations.
+ */
+class KDMQTT_API IMqttClient
+{
+public:
+    enum class ConnectionState {
+        CONNECTING,
+        CONNECTED,
+        DISCONNECTING,
+        DISCONNECTED
+    };
+
+    enum class SubscriptionState {
+        SUBSCRIBING,
+        SUBSCRIBED,
+        UNSUBSCRIBING,
+        UNSUBSCRIBED
+    };
+
+    KDBindings::Property<ConnectionState> connectionState{ ConnectionState::DISCONNECTED };
+    KDBindings::Property<SubscriptionState> subscriptionState{ SubscriptionState::UNSUBSCRIBED };
+
+    KDBindings::Property<std::vector<std::string>> subscriptions{};
+
+    KDBindings::Signal<int /*msgId*/> msgPublished;
+    KDBindings::Signal<const mosquitto_message * /*msg*/> msgReceived;
+
+    KDBindings::Signal<> error;
+
+    virtual int setTls(const File &cafile) = 0;
+    virtual int setUsernameAndPassword(const std::string &username, const std::string &password) = 0;
+    virtual int setWill(const std::string &topic, int payloadlen = 0, const void *payload = nullptr, int qos = 0, bool retain = false) = 0;
+
+    virtual int connect(const Url &host, int port = c_defaultPort, int keepalive = c_defaultKeepAliveSeconds) = 0;
+    virtual int disconnect() = 0;
+
+    virtual int publish(int *msgId, const char *topic, int payloadlen = 0, const void *payload = nullptr, int qos = 0, bool retain = false) = 0;
+
+    virtual int subscribe(const char *pattern, int qos = 0) = 0;
+    virtual int unsubscribe(const char *pattern) = 0;
+};
+
+/*
+ * Class: MqttClient
+ *
+ * This class exposes mosquitto client functions to
+ * the application / business logic.
+ */
+class KDMQTT_API MqttClient : public IMqttClient
+{
+    friend class MqttUnitTestHarness;
+
+public:
+    MqttClient(const std::string &clientId, bool cleanSession = true, bool verbose = false);
+    ~MqttClient() = default;
+
+    int setTls(const File &cafile) override;
+    int setUsernameAndPassword(const std::string &username, const std::string &password) override;
+    int setWill(const std::string &topic, int payloadlen = 0, const void *payload = nullptr, int qos = 0, bool retain = false) override;
+
+    int connect(const Url &host, int port = c_defaultPort, int keepalive = c_defaultKeepAliveSeconds) override;
+    int disconnect() override;
+
+    int publish(int *msgId, const char *topic, int payloadlen = 0, const void *payload = nullptr, int qos = 0, bool retain = false) override;
+
+    int subscribe(const char *pattern, int qos = 0) override;
+    int unsubscribe(const char *pattern) override;
+
+private:
+    bool m_verbose;
+
+    /*
+     * Mosquitto client event handlers
+     */
+    void onConnected(int connackCode);
+    void onDisconnected(int reasonCode);
+    void onPublished(int msgId);
+    void onMessage(const struct mosquitto_message *message);
```

Comment:
I'd rather avoid leaking the underlying library type. That will make us dependent on the implementation. We should copy the payload into our `KDUtils::ByteArray`.

That will be particularly wrong if one day we have event loops per thread and user might receive the data on a different thread. We'd pack the pointer which will be gone because mosquitto will internally already freed it.